example/src/main/java/io/netty/example/stomp/websocket/StompChatHandler.java

```diff
@@ -0,0 +1,191 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.example.stomp.websocket;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.DecoderResult;
+import io.netty.handler.codec.stomp.DefaultStompFrame;
+import io.netty.handler.codec.stomp.StompCommand;
+import io.netty.handler.codec.stomp.StompFrame;
+import io.netty.util.CharsetUtil;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static io.netty.handler.codec.stomp.StompHeaders.*;
+
+@Sharable
+public class StompChatHandler extends SimpleChannelInboundHandler<StompFrame> {
+
+    private final ConcurrentMap<String, Set<StompSubscription>> chatDestinations =
+            new ConcurrentHashMap<String, Set<StompSubscription>>();
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, StompFrame inboundFrame) throws Exception {
+        DecoderResult decoderResult = inboundFrame.decoderResult();
+        if (decoderResult.isFailure()) {
+            sendErrorFrame("rejected frame", decoderResult.toString(), ctx);
+            return;
+        }
+
+        switch (inboundFrame.command()) {
+        case STOMP:
+        case CONNECT:
+            onConnect(ctx, inboundFrame);
+            break;
+        case SUBSCRIBE:
+            onSubscribe(ctx, inboundFrame);
+            break;
+        case SEND:
+            onSend(ctx, inboundFrame);
+            break;
+        case UNSUBSCRIBE:
+            onUnsubscribe(ctx, inboundFrame);
+            break;
+        case DISCONNECT:
+            onDisconnect(ctx, inboundFrame);
+            break;
+        default:
+            sendErrorFrame("unsupported command",
+                           "Received unsupported command " + inboundFrame.command(), ctx);
+        }
+    }
+
+    private void onSubscribe(ChannelHandlerContext ctx, StompFrame inboundFrame) {
+        String destination = inboundFrame.headers().getAsString(DESTINATION);
+        String subscriptionId = inboundFrame.headers().getAsString(ID);
+
+        if (destination == null || subscriptionId == null) {
+            sendErrorFrame("missed header", "Required 'destination' or 'id' header missed", ctx);
+            return;
+        }
+
+        Set<StompSubscription> subscriptions = chatDestinations.get(destination);
+        if (subscriptions == null) {
+            subscriptions = new HashSet<StompSubscription>();
+            if (chatDestinations.putIfAbsent(destination, subscriptions) != null) {
+                subscriptions = chatDestinations.get(destination);
```

Comment:
ni: there is no need to call get again... `putIfAbsent` will return the subscription already so just store it in another var and re-assign

Reply:
Done